### PR TITLE
Improve performance of determining max height of receiver constraints

### PR DESCRIPTION
**Pull request by bbaldino**
When doing large call testing, we noticed a very hot call path where we figure out what the maximum height of a set of constraints is.  This change keeps track of the max height as constraints are added and removed from the map so we don't have to walk the entire thing to find it.

Note: `AbstractEndpoint#receiverVideoConstraintsChanged` should probably take a some structure which has max values for various constraints, but currently we only look at maxHeight so rather than define that new class now, I changed the function to just take `int maxHeight`.  It won't be difficult to expand that later when we use other constraints.

### CODE REVIEW

File: jvb/src/main/kotlin/org/jitsi/videobridge/cc/allocation/ReceiverConstraintsMap.kt [hunk 1]
```diff
@@ -39,9 +39,22 @@ class ReceiverConstraintsMap {
...
                 if (removed.maxHeight == maxHeight) {
-                    maxHeight = map.values.map { it.maxHeight }.max() ?: 0
+                    maxHeight = findNewMax(maxHeight)
                 }
             }
         }
     }
+
+    // Note: this method must be called with the lock held
+    private fun findNewMax(oldMax: Int): Int {
```
Comment (JonathanLennox) [inline jvb/src/main/kotlin/org/jitsi/videobridge/cc/allocation/ReceiverConstraintsMap.kt:49]: Maybe name this so it's clear it's only applicable for `remove`?

**Comment by JonathanLennox** [comment]
The alternative to defining another new object for `AbstractEndpoint#receiverVideoConstraintsChanged` would be to pass it the new `ReceiverConstraintsMap` itself.

File: jvb/src/main/kotlin/org/jitsi/videobridge/cc/allocation/ReceiverConstraintsMap.kt (hunk 1 above)
Comment (bbaldino) [reply jvb/src/main/kotlin/org/jitsi/videobridge/cc/allocation/ReceiverConstraintsMap.kt:49]: Well, technically it doesn't care when it's used.  I would've done it in `remove` itself but with the early returns it's tricky.

**Comment by bbaldino** [comment]
> The alternative to defining another new object for `AbstractEndpoint#receiverVideoConstraintsChanged` would be to pass it the new `ReceiverConstraintsMap` itself.

Yeah, my thought was to just take what was needed.  Even if/when we do need more things, it might be nice to minimize the places where `ReceiverConstraintsMap` is passed, though, to minimize dependencies on that API.

File: jvb/src/main/kotlin/org/jitsi/videobridge/cc/allocation/ReceiverConstraintsMap.kt (hunk 1 above)
Comment (JonathanLennox) [reply jvb/src/main/kotlin/org/jitsi/videobridge/cc/allocation/ReceiverConstraintsMap.kt:51]: Well, it's not correct for `add`, but fortunately `add`'s calculation is easy.

Comment (bbaldino) [reply jvb/src/main/kotlin/org/jitsi/videobridge/cc/allocation/ReceiverConstraintsMap.kt:51]: Yeah, actually you're right.  Will rename.

File: jvb/src/main/kotlin/org/jitsi/videobridge/cc/allocation/ReceiverConstraintsMap.kt [hunk 2]
```diff
@@ -57,4 +59,9 @@ class ReceiverConstraintsMap {
         }
         return maxSeen
     }
+
+    fun getDebugState() = OrderedJsonObject().apply {
+        put("maxHeight", maxHeight)
+        put("constraints", map)
```
Comment (JonathanLennox) [inline jvb/src/main/kotlin/org/jitsi/videobridge/cc/allocation/ReceiverConstraintsMap.kt:65]: Does this leak the map object outside the synchronization?

Comment (bbaldino) [reply jvb/src/main/kotlin/org/jitsi/videobridge/cc/allocation/ReceiverConstraintsMap.kt:65]: Technically, yes.  I can do `map.toMap()` to make a copy there.

**Review by JonathanLennox — APPROVED** [review]

**Pull request merged**